2_run/src/run_glm3.R

```diff
@@ -0,0 +1,150 @@
+  on.exit(unlink(sim_lake_dir, recursive = TRUE))
+  
+  # Define time period begin and end dates
+  times <- strsplit(time_period,'_')[[1]]
+  time_period_begin <- sprintf('%s-01-01', times[1])
+  time_period_end <- sprintf('%s-12-31', times[2])
+  
+  # Read in meteo data, add burn in and burn out and save to sim_lake_dir
+  meteo_data <- munge_meteo_w_burn_in_out(meteo_xwalk, time_period_begin, time_period_end, 
+                                          burn_in = 300, burn_out = 190)
+  sim_meteo_filename <- 'meteo_fl.csv'
+  readr::write_csv(meteo_data, file.path(sim_lake_dir, sim_meteo_filename))
+  
+  # Define simulation start and stop dates using munged meteo data
+  sim_start <- format(meteo_data[1,]$time, "%Y-%m-%d")
+  sim_stop <- format(tail(meteo_data, 1L)$time, "%Y-%m-%d")
+  
+  # write nml file, specifying meteo file and start and stop dates:
+  nml_obj <- nml_objs[[lake_id]]
+  nml_obj <- set_nml(nml_obj, arg_list = list(nsave = 24, 
+                                              meteo_fl = sim_meteo_filename,
+                                              sim_name = sprintf('%s_%s_%s', lake_id, gcm, time_period),
+                                              start = sim_start,
+                                              stop = sim_stop))
+  nml_obj$light$Kw_file <- NULL # use static value instead of relying on a file
+  glmtools::write_nml(nml_obj, file.path(sim_lake_dir, 'glm3.nml'))
+  
+  # Set up export tibble
+  export_fl <- sprintf(export_fl_template, lake_id, gcm, time_period)
+  export_tibble <- tibble(
+    run_date = Sys.time(),
+    lake_id = lake_id,
+    gcm = gcm,
+    time_period = time_period,
+    export_fl = export_fl)
+  
+  # for each model run, try running the model up to 5 times
+  # if model run works, report how long that model took to run
+  # and extract output and save to feather export file
+  # return tibble with export file name and hash
+  # if model run fails after 5 attempts, return
+  # tibble indicating that it failed and giving the code
+  tryCatch(
+    {
+      retry::retry(
+        {glm_time <- system.time({glm_code <- GLM3r::run_glm(sim_lake_dir, verbose = FALSE)})[['elapsed']]},
+        until=function(val, cnd) glm_code==0,
+        max_tries = 5)
+      
+      # make sure glm did succeed
+      if(glm_code != 0) stop()
+      
+      # extract output
+      extract_glm_output(sim_lake_dir, nml_obj, export_fl)
+      
+      # Add glm run information to export tibble
+      export_tibble <- export_tibble %>%
+        mutate(export_hash = tools::md5sum(export_fl),
+               glm_time_s = glm_time,
+               glm_success = TRUE,
+               glm_code = glm_code)
+      return(export_tibble)
+    },
+    error = function(e) {
+      glm_time <- system.time({glm_code <- GLM3r::run_glm(sim_lake_dir, verbose = FALSE)})[['elapsed']]
+      
+      # Make sure glm did indeed fail
+      if(glm_code == 0) stop()
```

Comment:
Ah, yes. I realize now that I was mistakenly thinking that I needed to run glm one final time so that I could pull the code returned by `GLM3r::run_glm()` for the failed run, but I realize now that code would already be stored in `glm_code`, so I can just delete both the 6th glm run and the catch I added in case `glm_code == 0`, which would fully error out the build, as you noticed. I think it's better to omit that second check, given that the try block should only trigger the error function if the `glm_code` is not 0.

Reply:
^ worth noting that I think this pattern popped up in code Alison had written and @hcorson-dosch was trying to emulate here. But, agree, that it can go.

Reply:
I see. I don't think Alison's code had a `stop()` within the error block and that was the part of this where I wasn't sure what happens. 
I think in that other repo, the second run of the model was to capture the verbose failure of the model, which looks like this:
```
Reading config from glm3.nml
No WQ config
values for crest_elev not provided, assuming max elevation, H[bsn]
VolAtCrest= 494638433.64669, and Max Lake Vol= 494638433.64669 (m3)
Sediment section present, simulating sediment heating
Cannot find file "met_data_hourly.csv"
Failed to open 'met_data_hourly.csv'
       ----------------------------------------------------
       |  General Lake Model (GLM)   Version 3.0.5        |
       ----------------------------------------------------
glm built using gcc version 4.2.1
nDays 730 timestep 3600.000000
Maximum lake depth is 24.993600
Depth where flow will occur over the crest is 24.993600
*starting sediment =    5.00000
*starting sediment =    0.10000
*sed_temp_mean =    7.28664
*sed_temp_mean =   17.40340
```
whereas the error code is `1`

Reply:
Yup you're right, Jordan. That `stop()` addition was mine, after I pulled in Alison's tryCatch code which used the error function to make a second call to `GLM3r::run_glm()` with verbose == TRUE. It's gone now.